src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -651,57 +625,52 @@ List<TelemetryEvent> ToTelemetryPackageList(List<Tuple<string, string>> packages
+            INuGetUI uiService,
+            CancellationToken cancellationToken)
         {
-            var potentialProjects = new List<NuGetProject>();
-
-            // check if project suppports <PackageReference> items.
-            // otherwise don't show format selector dialog for this project
-            var capableProjects = uiService
-                .Projects
-                .Where(project =>
-                    project.ProjectStyle == ProjectModel.ProjectStyle.PackagesConfig &&
-                    project.ProjectServices.Capabilities.SupportsPackageReferences);
-
-            // get all packages.config based projects with no installed packages
-            foreach (var project in capableProjects)
-            {
-                var installedPackages = await project.GetInstalledPackagesAsync(token);
+            IReadOnlyCollection<string> projectGuids = uiService.Projects.Select(project => project.ProjectId).ToArray();
+            IReadOnlyCollection<IProjectContextInfo> upgradeableProjects = await projectUpgrader.GetUpgradeableProjectsAsync(
+                projectGuids,
+                cancellationToken);
 
-                if (!installedPackages.Any())
-                {
-                    potentialProjects.Add(project);
-                }
-            }
-            
             // only show this dialog if there are any new project(s) with no installed packages.
-            if (potentialProjects.Count > 0)
+            if (upgradeableProjects.Count > 0)
             {
                 var packageManagementFormat = new PackageManagementFormat(uiService.Settings);
+
                 if (!packageManagementFormat.Enabled)
                 {
                     // user disabled this prompt either through Tools->options or previous interaction of this dialog.
                     // now check for default package format, if its set to PackageReference then update the project.
                     if (packageManagementFormat.SelectedPackageManagementFormat == 1)
                     {
-                        await uiService.UpdateNuGetProjectToPackageRef(potentialProjects);
+                        await uiService.UpgradeProjectsToPackageReferenceAsync(upgradeableProjects);
                     }
 
                     return true;
                 }
 
-                packageManagementFormat.ProjectNames = potentialProjects
-                    .Select(project => project.GetMetadata<string>(NuGetProjectMetadataKeys.Name))
-                    .OrderBy(name => name, StringComparer.OrdinalIgnoreCase).ToList();
+                Task<string>[] tasks = upgradeableProjects
+                    .Select(project => project.GetMetadataAsync<string>(NuGetProjectMetadataKeys.Name, cancellationToken).AsTask())
+                    .ToArray();
+
+                await Task.WhenAll(tasks);
+
+                packageManagementFormat.ProjectNames = tasks.Select(task => task.Result)
```

Comment:
Any reason you don't get the value from await Task.WhenAll instead of using Linq.Select(task.result?